File: Data Structures & Algorithms/course-schedule-iv/submission-7.cpp
class Solution {
    vector<vector<int>> adj;
    unordered_map<int, unordered_set<int>> prereqMap;

    void topo(int node){

        if (prereqMap.count(node)){
            return;
        }

        for (auto nei: adj[node]){
            topo(nei);
            prereqMap[node].insert(prereqMap[nei].begin(), prereqMap[nei].end());
        }
        prereqMap[node].insert(node);
    }

public:
    vector<bool> checkIfPrerequisite(int numCourses, vector<vector<int>>& prerequisites, vector<vector<int>>& queries) {
        adj.assign(numCourses, vector<int>());
        for (vector<int>& pre: prerequisites){
            adj[pre[1]].push_back(pre[0]);
        }

        for (int i = 0; i < numCourses; i++){
            topo(i);
        }

        vector<bool> res;

        for (auto& query: queries){
            res.push_back(prereqMap[query[1]].count(query[0]));
        }
        
        return res;
    }
};

File: Data Structures & Algorithms/course-schedule/submission-7.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        adjList = {}

        if len(prerequisites) == 0:
            return True

        for src, dst in prerequisites:
            if src not in adjList:
                adjList[src] = []
            if dst not in adjList:
                adjList[dst] = []
            adjList[src].append(dst)

        visit = set()
        def dfs(node):
            if node in visit:
                return False

            if adjList[node] == []:
                return True

            visit.add(node)

            for nei in adjList[node]:
                if not dfs(nei):
                    return False
                
            visit.remove(node)
            adjList[node] = []

            return True

        for key in adjList:
            if not dfs(key):
                return False

        return True